In [1]:
import pandas as pd
from collections import defaultdict
import networkx as nx
import pickle
import collections
import math
from tqdm import tqdm

In [2]:
source_destination=pd.read_csv('wiki-topcats-reduced.txt',sep="\t",header=None,names=["source","destination"])

In [3]:
source_destination.head()

source  destination
0      52       401135
1      52      1069112
2      52      1163551
3      62        12162
4      62       167659

In [4]:
source_destination_dict = pickle.load(open('source_destination_dict.p', 'rb'))

In [5]:
source_destination_dict

defaultdict(list,
            {52: [401135, 1069112, 1163551],
             62: [12162,
              167659,
              279122,
              1089199,
              1354553,
              1400636,
              1403619,
              1537692,
              1544420],
             64: [64873],
             66: [279122, 1163290],
             74: [279122],
             95: [1179478, 1185516, 1185519, 1230865],
             103: [107],
             104: [107, 1174251, 1174302],
             105: [1174967],
             107: [104,
              142167,
              657930,
              1013995,
              1061780,
              1174302,
              1179210,
              1181401],
             108: [104, 107, 1059989, 1062426, 1161925, 1169534],
             112: [107],
             113: [279122, 1185127],
             122: [1174967],
             126: [279122],
             133: [279122, 825464],
             134: [541222,
              1060414,
              1061485,
          

In [6]:
source_destination_tuples=[(row["source"],row["destination"]) for idx,row in source_destination.iterrows()]

In [8]:
source_destination_tuples

[(52, 401135),
 (52, 1069112),
 (52, 1163551),
 (62, 12162),
 (62, 167659),
 (62, 279122),
 (62, 1089199),
 (62, 1354553),
 (62, 1400636),
 (62, 1403619),
 (62, 1537692),
 (62, 1544420),
 (64, 64873),
 (66, 279122),
 (66, 1163290),
 (74, 279122),
 (95, 1179478),
 (95, 1185516),
 (95, 1185519),
 (95, 1230865),
 (103, 107),
 (104, 107),
 (104, 1174251),
 (104, 1174302),
 (105, 1174967),
 (107, 104),
 (107, 142167),
 (107, 657930),
 (107, 1013995),
 (107, 1061780),
 (107, 1174302),
 (107, 1179210),
 (107, 1181401),
 (108, 104),
 (108, 107),
 (108, 1059989),
 (108, 1062426),
 (108, 1161925),
 (108, 1169534),
 (112, 107),
 (113, 279122),
 (113, 1185127),
 (122, 1174967),
 (126, 279122),
 (133, 279122),
 (133, 825464),
 (134, 541222),
 (134, 1060414),
 (134, 1061485),
 (134, 1062226),
 (134, 1163610),
 (134, 1163783),
 (134, 1265505),
 (137, 136),
 (137, 1163286),
 (137, 1163712),
 (137, 1224802),
 (153, 154),
 (153, 260842),
 (153, 725249),
 (153, 744272),
 (153, 1064807),
 (153, 1065282),


In [9]:
#The number of nodes
Gtup = nx.Graph()
Gtup.add_edges_from(source_destination_tuples)
print("Number of nodes:",len(Gtup.nodes()),"\nNumber of edges:",len(Gtup.edges()))

Number of nodes: 461193 
Number of edges: 2174451


In [9]:
nx.info(Gtup).split("\n")[4]

'Average degree:   9.4297'

In [12]:
nx.density(Gtup)

2.0446319786157265e-05

In [13]:
first_key=next(iter(source_destination_dict))

(first_key,source_destination_dict[first_key])

(52, [401135, 1069112, 1163551])

In [14]:
nx.is_directed(Gtup)

False

In [15]:
indicatorD=False
first_key=next(iter(source_destination_dict))
first_key_tup=(first_key,source_destination_dict[first_key])
for dest in first_key_tup[1]:#check when destination is source
    if first_key not in source_destination_dict[dest]:
        indicatorD=True
if indicatorD==True:
    print("Graph is directed")
else:
    print("Graph is undirected")

Graph is directed


In [10]:
Gtup_directed = nx.DiGraph()

In [11]:
Gtup_directed.add_edges_from(source_destination_tuples)
print("Number of nodes:",len(Gtup_directed.nodes()),"\nNumber of edges:",len(Gtup_directed.edges()))

Number of nodes: 461193 
Number of edges: 2645247


In [18]:
nx.is_directed(Gtup_directed)

True

In [12]:
categories=pd.read_csv('wiki-topcats-categories.txt',sep="\n",header=None)

In [13]:
def num_cat(row):
    cat=row
    return len(cat.split("; ")[1].split(" "))

In [14]:
categories=categories[categories.iloc[:,0].apply(lambda x:True if num_cat(x)>3500 else False)]

In [15]:
categories

0
868    Category:English_footballers; 22860 28411 2896...
869    Category:The_Football_League_players; 14003 23...
876    Category:Association_football_forwards; 26876 ...
898    Category:Association_football_goalkeepers; 269...
900    Category:Association_football_midfielders; 140...
949    Category:Association_football_defenders; 15217...
1047   Category:Living_people; 52 62 64 66 70 74 79 9...
1881   Category:Year_of_birth_unknown; 3335 10527 150...
2791   Category:Harvard_University_alumni; 77 1013 12...
4160   Category:Major_League_Baseball_pitchers; 79 24...
5619   Category:Members_of_the_United_Kingdom_Parliam...
6067   Category:Indian_films; 1308 29286 53565 70274 ...
6795   Category:Year_of_death_missing; 98 126 227 182...
6893   Category:English_cricketers; 845 2658 3277 228...
7628   Category:Year_of_birth_missing_(living_people)...
7889   Category:Rivers_of_Romania; 72111 72112 72113 ...
8732   Category:Main_Belt_asteroids; 13184 13992 2227...
8733   Category:Asteroids_named_for_people; 13184 139...
10062  Category:English-language_albums; 185 186 4052...
10139  Category:English_television_actors; 32782 4033...
10251  Category:British_films; 214 1084 23768 28836 3...
10555  Category:English-language_films; 55 134 153 21...
10564  Category:American_films; 134 153 173 1083 1087...
10839  Category:Fellows_of_the_Royal_Society; 1670 16...
11640  Category:People_from_New_York_City; 137 980 16...
11641  Category:American_Jews; 167 1101 1286 2466 507...
11648  Category:American_television_actors; 174 938 9...
11649  Category:American_film_actors; 174 980 1088 10...
11992  Category:Debut_albums; 95 108 112 113 190 254 ...
12631  Category:Black-and-white_films; 134 16130 1976...
12941  Category:Year_of_birth_missing; 147 171 1056 1...
12966  Category:Place_of_birth_missing_(living_people...
13838  Category:Article_Feedback_Pilot; 4500 4924 525...
13996  Category:American_military_personnel_of_World_...
16956  Category:Windows_games; 14095 18840 19757 2284...

In [16]:
categories_dict=defaultdict(list)
for idx in categories.index:
    cat_and_values=categories[0].loc[idx].split("; ")
    cat_name=cat_and_values[0].split(":")[1]
    categories_dict[cat_name]=list(map(int,cat_and_values[1].split(" ")))

In [58]:
categories_dict['English_television_actors']


[32782,
 40338,
 40566,
 53495,
 70173,
 72636,
 75376,
 76604,
 83294,
 84078,
 85090,
 92536,
 94198,
 94537,
 96274,
 109867,
 109890,
 113194,
 113259,
 131855,
 131913,
 134497,
 137747,
 142496,
 145391,
 146728,
 146750,
 151481,
 151482,
 151506,
 152932,
 156192,
 157543,
 158111,
 158448,
 164301,
 165140,
 166620,
 167466,
 172559,
 175252,
 184837,
 205296,
 208941,
 209040,
 210280,
 210464,
 210755,
 212777,
 220519,
 230647,
 247467,
 247476,
 247485,
 255982,
 255983,
 256816,
 290953,
 332118,
 340938,
 343202,
 343281,
 343408,
 343409,
 348794,
 350088,
 356680,
 369362,
 369424,
 373339,
 376202,
 376209,
 391372,
 392767,
 400868,
 401893,
 401983,
 402670,
 402899,
 418183,
 419816,
 423988,
 428273,
 430640,
 430653,
 430662,
 430673,
 430675,
 430686,
 430720,
 430748,
 430749,
 430782,
 430783,
 430787,
 431375,
 431603,
 436006,
 436009,
 438370,
 449340,
 451046,
 453686,
 466551,
 467110,
 469565,
 480173,
 481038,
 499468,
 499543,
 517246,
 526313,
 526315

In [17]:
for key, values in categories_dict.items():
    print(key, len(values))

English_footballers 9237
The_Football_League_players 9467
Association_football_forwards 6959
Association_football_goalkeepers 3997
Association_football_midfielders 8270
Association_football_defenders 6668
Living_people 418223
Year_of_birth_unknown 3760
Harvard_University_alumni 6154
Major_League_Baseball_pitchers 6580
Members_of_the_United_Kingdom_Parliament_for_English_constituencies 6546
Indian_films 5913
Year_of_death_missing 7851
English_cricketers 3813
Year_of_birth_missing_(living_people) 34721
Rivers_of_Romania 7729
Main_Belt_asteroids 13704
Asteroids_named_for_people 5701
English-language_albums 4853
English_television_actors 3501
British_films 4551
English-language_films 22699
American_films 15302
Fellows_of_the_Royal_Society 3697
People_from_New_York_City 4888
American_Jews 3542
American_television_actors 11661
American_film_actors 13938
Debut_albums 8401
Black-and-white_films 12174
Year_of_birth_missing 7237
Place_of_birth_missing_(living_people) 6767
Article_Feedback_Pilot 

We can see that the categories with minus articles is 'English_television_actors' (3501). So we decide to deal with this category. This category'll be our input category

In [69]:
len(categories_dict['English_television_actors'])

3501

In [18]:
from collections import deque


In [49]:
def bfs_shortest_path(graph, start):
    explored = set()
    queue = [start]
    shortest_path_dict = {}
    counter_level = 0
    to_visit_list = []
    while queue:
        for node in queue:
            if node not in explored:
                shortest_path_dict[node] = counter_level
                neighbours = graph[node]
                explored.add(node)
                to_visit_list.extend(neighbours)
        queue = []
        queue.extend(to_visit_list)
        to_visit_list = []
        counter_level += 1
    return shortest_path_dict

In [101]:
a = bfs_shortest_path(Gtup_directed, categories_dict['English_television_actors'][0])

In [102]:
a

{32782: 0,
 351118: 1,
 1061148: 1,
 1061960: 1,
 1061963: 1,
 1062082: 1,
 1063756: 1,
 1065778: 1,
 216813: 2,
 350752: 2,
 350894: 2,
 350927: 2,
 351001: 2,
 351128: 2,
 543241: 2,
 1036190: 2,
 1058794: 2,
 1223412: 2,
 1394526: 2,
 430722: 2,
 1031462: 2,
 1054907: 2,
 1055543: 2,
 1056766: 2,
 1058223: 2,
 1059302: 2,
 1060200: 2,
 1060725: 2,
 1060736: 2,
 1060737: 2,
 1060739: 2,
 1060891: 2,
 1061149: 2,
 1061150: 2,
 1061664: 2,
 1061824: 2,
 1061825: 2,
 1061832: 2,
 1061848: 2,
 1061861: 2,
 1061885: 2,
 1061922: 2,
 1061939: 2,
 1062050: 2,
 1062484: 2,
 1062688: 2,
 1062748: 2,
 1064363: 2,
 1065611: 2,
 1066978: 2,
 56387: 2,
 272763: 2,
 279122: 2,
 324958: 2,
 325554: 2,
 338810: 2,
 493443: 2,
 742511: 2,
 744381: 2,
 751230: 2,
 817478: 2,
 824119: 2,
 1028775: 2,
 1028948: 2,
 1045097: 2,
 1056565: 2,
 1058108: 2,
 1058661: 2,
 1058773: 2,
 1058843: 2,
 1058933: 2,
 1059204: 2,
 1059373: 2,
 1059441: 2,
 1059645: 2,
 1059646: 2,
 1059647: 2,
 1059674: 2,
 1059741: 

In [113]:
shortest_path_dictionary_final = {}
for article in tqdm(range(0,len(categories_dict['English_television_actors']))):
    try:
        shortest_path_dictionary_final.update(bfs_shortest_path(Gtup_directed, categories_dict['English_television_actors'][article]))   
    except:
        shortest_path_dictionary_final[article] = -1






  0%|          | 0/3501 [00:00<?, ?it/s]




  0%|          | 1/3501 [00:01<59:48,  1.03s/it]




  0%|          | 2/3501 [00:02<1:00:27,  1.04s/it]




  0%|          | 3/3501 [00:03<1:01:08,  1.05s/it]




  0%|          | 4/3501 [00:04<1:00:59,  1.05s/it]




  0%|          | 6/3501 [00:05<51:54,  1.12it/s]  




  0%|          | 9/3501 [00:06<42:38,  1.36it/s]




  0%|          | 10/3501 [00:07<48:06,  1.21it/s]




  0%|          | 11/3501 [00:08<51:56,  1.12it/s]




  0%|          | 12/3501 [00:09<55:25,  1.05it/s]




  0%|          | 14/3501 [00:10<48:18,  1.20it/s]




  0%|          | 16/3501 [00:11<43:03,  1.35it/s]




  1%|          | 18/3501 [00:12<39:14,  1.48it/s]




  1%|          | 19/3501 [00:13<45:24,  1.28it/s]




  1%|          | 20/3501 [00:14<49:43,  1.17it/s]




  1%|          | 21/3501 [00:15<53:17,  1.09it/s]




  1%|          | 22/3501 [00:16<56:51,  1.02it/s]




  1%|          | 23/3501 [00:17<57:12,  1.01it/s]




  1%|          | 24/3501 [00:1

 10%|█         | 366/3501 [05:09<43:06,  1.21it/s]




 10%|█         | 367/3501 [05:10<47:03,  1.11it/s]




 11%|█         | 369/3501 [05:11<41:30,  1.26it/s]




 11%|█         | 370/3501 [05:12<51:21,  1.02it/s]




 11%|█         | 371/3501 [05:14<1:01:48,  1.18s/it]




 11%|█         | 372/3501 [05:15<1:02:11,  1.19s/it]




 11%|█         | 374/3501 [05:16<51:45,  1.01it/s]  




 11%|█         | 377/3501 [05:17<41:37,  1.25it/s]




 11%|█         | 379/3501 [05:19<37:38,  1.38it/s]




 11%|█         | 380/3501 [05:20<43:43,  1.19it/s]




 11%|█         | 381/3501 [05:21<47:46,  1.09it/s]




 11%|█         | 382/3501 [05:22<50:38,  1.03it/s]




 11%|█         | 383/3501 [05:23<52:18,  1.01s/it]




 11%|█         | 384/3501 [05:24<52:37,  1.01s/it]




 11%|█         | 385/3501 [05:25<53:44,  1.03s/it]




 11%|█         | 386/3501 [05:26<56:09,  1.08s/it]




 11%|█         | 387/3501 [05:27<56:54,  1.10s/it]




 11%|█         | 388/3501 [05:28<56:38,  1.09s/it]




 11%

 20%|█▉        | 699/3501 [10:21<48:48,  1.05s/it]




 20%|█▉        | 700/3501 [10:22<48:22,  1.04s/it]




 20%|██        | 701/3501 [10:23<48:17,  1.03s/it]




 20%|██        | 702/3501 [10:24<48:13,  1.03s/it]




 20%|██        | 703/3501 [10:25<47:56,  1.03s/it]




 20%|██        | 704/3501 [10:26<48:24,  1.04s/it]




 20%|██        | 705/3501 [10:27<49:47,  1.07s/it]




 20%|██        | 706/3501 [10:28<49:56,  1.07s/it]




 20%|██        | 708/3501 [10:29<42:08,  1.10it/s]




 20%|██        | 709/3501 [10:30<44:04,  1.06it/s]




 20%|██        | 710/3501 [10:31<45:01,  1.03it/s]




 20%|██        | 711/3501 [10:32<45:47,  1.02it/s]




 20%|██        | 712/3501 [10:34<46:18,  1.00it/s]




 20%|██        | 713/3501 [10:35<46:33,  1.00s/it]




 20%|██        | 714/3501 [10:36<46:43,  1.01s/it]




 20%|██        | 715/3501 [10:37<46:52,  1.01s/it]




 20%|██        | 716/3501 [10:38<47:18,  1.02s/it]




 20%|██        | 717/3501 [10:39<47:17,  1.02s/it]




 21%|██   

 24%|██▍       | 853/3501 [12:54<44:32,  1.01s/it]




 24%|██▍       | 854/3501 [12:55<44:37,  1.01s/it]




 24%|██▍       | 855/3501 [12:56<44:14,  1.00s/it]




 24%|██▍       | 856/3501 [12:57<44:22,  1.01s/it]




 24%|██▍       | 857/3501 [12:58<44:15,  1.00s/it]




 25%|██▍       | 858/3501 [12:59<44:12,  1.00s/it]




 25%|██▍       | 859/3501 [13:00<44:21,  1.01s/it]




 25%|██▍       | 860/3501 [13:01<44:13,  1.00s/it]




 25%|██▍       | 861/3501 [13:02<44:03,  1.00s/it]




 25%|██▍       | 862/3501 [13:03<44:20,  1.01s/it]




 25%|██▍       | 863/3501 [13:04<44:11,  1.01s/it]




 25%|██▍       | 864/3501 [13:05<43:54,  1.00it/s]




 25%|██▍       | 865/3501 [13:06<44:06,  1.00s/it]




 25%|██▍       | 866/3501 [13:07<44:02,  1.00s/it]




 25%|██▍       | 867/3501 [13:08<43:59,  1.00s/it]




 25%|██▍       | 868/3501 [13:09<43:45,  1.00it/s]




 25%|██▍       | 869/3501 [13:10<43:32,  1.01it/s]




 25%|██▍       | 870/3501 [13:11<43:37,  1.01it/s]




 25%|██▍  

 33%|███▎      | 1153/3501 [17:52<39:09,  1.00s/it]




 33%|███▎      | 1155/3501 [17:53<33:21,  1.17it/s]




 33%|███▎      | 1156/3501 [17:54<34:52,  1.12it/s]




 33%|███▎      | 1157/3501 [17:55<36:06,  1.08it/s]




 33%|███▎      | 1158/3501 [17:56<37:09,  1.05it/s]




 33%|███▎      | 1159/3501 [17:57<37:54,  1.03it/s]




 33%|███▎      | 1160/3501 [17:58<38:38,  1.01it/s]




 33%|███▎      | 1161/3501 [17:59<38:42,  1.01it/s]




 33%|███▎      | 1162/3501 [18:00<38:45,  1.01it/s]




 33%|███▎      | 1163/3501 [18:01<38:48,  1.00it/s]




 33%|███▎      | 1164/3501 [18:02<38:57,  1.00s/it]




 33%|███▎      | 1166/3501 [18:03<33:01,  1.18it/s]




 33%|███▎      | 1167/3501 [18:04<34:57,  1.11it/s]




 33%|███▎      | 1168/3501 [18:05<36:07,  1.08it/s]




 33%|███▎      | 1169/3501 [18:06<36:55,  1.05it/s]




 33%|███▎      | 1170/3501 [18:07<37:29,  1.04it/s]




 33%|███▎      | 1172/3501 [18:08<32:03,  1.21it/s]




 34%|███▎      | 1173/3501 [18:09<34:03,  1.14it

 42%|████▏     | 1458/3501 [22:51<34:25,  1.01s/it]




 42%|████▏     | 1459/3501 [22:52<34:29,  1.01s/it]




 42%|████▏     | 1460/3501 [22:53<34:19,  1.01s/it]




 42%|████▏     | 1461/3501 [22:54<34:12,  1.01s/it]




 42%|████▏     | 1462/3501 [22:55<34:07,  1.00s/it]




 42%|████▏     | 1463/3501 [22:56<33:55,  1.00it/s]




 42%|████▏     | 1464/3501 [22:57<33:46,  1.01it/s]




 42%|████▏     | 1465/3501 [22:58<33:48,  1.00it/s]




 42%|████▏     | 1466/3501 [22:59<33:49,  1.00it/s]




 42%|████▏     | 1467/3501 [23:00<33:30,  1.01it/s]




 42%|████▏     | 1468/3501 [23:01<33:37,  1.01it/s]




 42%|████▏     | 1469/3501 [23:02<33:40,  1.01it/s]




 42%|████▏     | 1470/3501 [23:03<33:25,  1.01it/s]




 42%|████▏     | 1471/3501 [23:04<33:21,  1.01it/s]




 42%|████▏     | 1472/3501 [23:05<34:36,  1.02s/it]




 42%|████▏     | 1473/3501 [23:06<34:24,  1.02s/it]




 42%|████▏     | 1474/3501 [23:07<34:25,  1.02s/it]




 42%|████▏     | 1475/3501 [23:08<34:58,  1.04s/

 50%|█████     | 1765/3501 [27:52<27:01,  1.07it/s]




 50%|█████     | 1766/3501 [27:53<27:34,  1.05it/s]




 50%|█████     | 1767/3501 [27:54<27:58,  1.03it/s]




 50%|█████     | 1768/3501 [27:55<28:13,  1.02it/s]




 51%|█████     | 1769/3501 [27:56<28:24,  1.02it/s]




 51%|█████     | 1770/3501 [27:57<28:40,  1.01it/s]




 51%|█████     | 1772/3501 [27:58<24:20,  1.18it/s]




 51%|█████     | 1773/3501 [27:59<25:46,  1.12it/s]




 51%|█████     | 1775/3501 [28:00<22:14,  1.29it/s]




 51%|█████     | 1776/3501 [28:01<24:01,  1.20it/s]




 51%|█████     | 1777/3501 [28:02<25:18,  1.14it/s]




 51%|█████     | 1778/3501 [28:03<26:11,  1.10it/s]




 51%|█████     | 1779/3501 [28:04<26:47,  1.07it/s]




 51%|█████     | 1780/3501 [28:05<27:21,  1.05it/s]




 51%|█████     | 1781/3501 [28:06<27:36,  1.04it/s]




 51%|█████     | 1782/3501 [28:07<27:53,  1.03it/s]




 51%|█████     | 1783/3501 [28:08<27:48,  1.03it/s]




 51%|█████     | 1784/3501 [28:09<27:55,  1.03it

 59%|█████▉    | 2062/3501 [33:34<29:44,  1.24s/it]




 59%|█████▉    | 2063/3501 [33:36<29:23,  1.23s/it]




 59%|█████▉    | 2064/3501 [33:37<29:29,  1.23s/it]




 59%|█████▉    | 2065/3501 [33:38<29:40,  1.24s/it]




 59%|█████▉    | 2066/3501 [33:39<30:03,  1.26s/it]




 59%|█████▉    | 2067/3501 [33:41<30:06,  1.26s/it]




 59%|█████▉    | 2068/3501 [33:42<29:57,  1.25s/it]




 59%|█████▉    | 2069/3501 [33:43<30:04,  1.26s/it]




 59%|█████▉    | 2070/3501 [33:44<29:52,  1.25s/it]




 59%|█████▉    | 2071/3501 [33:46<29:42,  1.25s/it]




 59%|█████▉    | 2072/3501 [33:47<29:51,  1.25s/it]




 59%|█████▉    | 2073/3501 [33:48<29:14,  1.23s/it]




 59%|█████▉    | 2074/3501 [33:49<28:43,  1.21s/it]




 59%|█████▉    | 2075/3501 [33:50<28:42,  1.21s/it]




 59%|█████▉    | 2076/3501 [33:52<28:31,  1.20s/it]




 59%|█████▉    | 2077/3501 [33:53<28:37,  1.21s/it]




 59%|█████▉    | 2078/3501 [33:54<28:49,  1.22s/it]




 59%|█████▉    | 2079/3501 [33:55<28:55,  1.22s/

 68%|██████▊   | 2368/3501 [39:09<19:47,  1.05s/it]




 68%|██████▊   | 2369/3501 [39:10<20:16,  1.07s/it]




 68%|██████▊   | 2370/3501 [39:12<20:15,  1.07s/it]




 68%|██████▊   | 2371/3501 [39:13<19:54,  1.06s/it]




 68%|██████▊   | 2374/3501 [39:14<15:45,  1.19it/s]




 68%|██████▊   | 2375/3501 [39:15<16:44,  1.12it/s]




 68%|██████▊   | 2376/3501 [39:16<17:24,  1.08it/s]




 68%|██████▊   | 2378/3501 [39:17<14:59,  1.25it/s]




 68%|██████▊   | 2379/3501 [39:18<16:06,  1.16it/s]




 68%|██████▊   | 2381/3501 [39:19<14:03,  1.33it/s]




 68%|██████▊   | 2384/3501 [39:20<11:39,  1.60it/s]




 68%|██████▊   | 2385/3501 [39:21<14:24,  1.29it/s]




 68%|██████▊   | 2386/3501 [39:22<16:15,  1.14it/s]




 68%|██████▊   | 2387/3501 [39:23<17:25,  1.07it/s]




 68%|██████▊   | 2388/3501 [39:24<18:16,  1.02it/s]




 68%|██████▊   | 2389/3501 [39:25<18:44,  1.01s/it]




 68%|██████▊   | 2390/3501 [39:26<18:58,  1.02s/it]




 68%|██████▊   | 2391/3501 [39:27<19:58,  1.08s/

 76%|███████▌  | 2668/3501 [44:17<16:52,  1.22s/it]




 76%|███████▌  | 2669/3501 [44:18<16:41,  1.20s/it]




 76%|███████▋  | 2670/3501 [44:19<16:04,  1.16s/it]




 76%|███████▋  | 2671/3501 [44:20<15:19,  1.11s/it]




 76%|███████▋  | 2672/3501 [44:21<14:55,  1.08s/it]




 76%|███████▋  | 2673/3501 [44:22<14:34,  1.06s/it]




 76%|███████▋  | 2674/3501 [44:23<14:46,  1.07s/it]




 76%|███████▋  | 2675/3501 [44:24<14:31,  1.05s/it]




 76%|███████▋  | 2676/3501 [44:25<14:27,  1.05s/it]




 76%|███████▋  | 2677/3501 [44:26<14:15,  1.04s/it]




 76%|███████▋  | 2678/3501 [44:27<14:16,  1.04s/it]




 77%|███████▋  | 2679/3501 [44:28<14:05,  1.03s/it]




 77%|███████▋  | 2680/3501 [44:29<15:01,  1.10s/it]




 77%|███████▋  | 2681/3501 [44:31<15:43,  1.15s/it]




 77%|███████▋  | 2682/3501 [44:32<15:32,  1.14s/it]




 77%|███████▋  | 2683/3501 [44:33<15:37,  1.15s/it]




 77%|███████▋  | 2684/3501 [44:34<15:04,  1.11s/it]




 77%|███████▋  | 2685/3501 [44:35<14:42,  1.08s/

 85%|████████▍ | 2971/3501 [49:24<09:39,  1.09s/it]




 85%|████████▍ | 2972/3501 [49:25<09:28,  1.07s/it]




 85%|████████▍ | 2973/3501 [49:26<09:18,  1.06s/it]




 85%|████████▍ | 2974/3501 [49:27<09:12,  1.05s/it]




 85%|████████▍ | 2975/3501 [49:28<09:09,  1.05s/it]




 85%|████████▌ | 2976/3501 [49:29<09:05,  1.04s/it]




 85%|████████▌ | 2977/3501 [49:30<09:03,  1.04s/it]




 85%|████████▌ | 2978/3501 [49:31<08:59,  1.03s/it]




 85%|████████▌ | 2979/3501 [49:32<08:56,  1.03s/it]




 85%|████████▌ | 2980/3501 [49:33<08:55,  1.03s/it]




 85%|████████▌ | 2981/3501 [49:34<08:54,  1.03s/it]




 85%|████████▌ | 2982/3501 [49:35<08:54,  1.03s/it]




 85%|████████▌ | 2983/3501 [49:36<08:51,  1.03s/it]




 85%|████████▌ | 2984/3501 [49:37<08:49,  1.02s/it]




 85%|████████▌ | 2985/3501 [49:38<08:50,  1.03s/it]




 85%|████████▌ | 2986/3501 [49:39<08:47,  1.03s/it]




 85%|████████▌ | 2987/3501 [49:40<08:46,  1.03s/it]




 85%|████████▌ | 2988/3501 [49:41<08:47,  1.03s/

 94%|█████████▍| 3303/3501 [54:33<03:20,  1.01s/it]




 94%|█████████▍| 3304/3501 [54:34<03:19,  1.01s/it]




 94%|█████████▍| 3305/3501 [54:35<03:18,  1.01s/it]




 94%|█████████▍| 3306/3501 [54:36<03:17,  1.01s/it]




 94%|█████████▍| 3307/3501 [54:37<03:16,  1.01s/it]




 94%|█████████▍| 3308/3501 [54:38<03:14,  1.01s/it]




 95%|█████████▍| 3309/3501 [54:39<03:13,  1.01s/it]




 95%|█████████▍| 3310/3501 [54:40<03:10,  1.00it/s]




 95%|█████████▍| 3311/3501 [54:41<03:10,  1.00s/it]




 95%|█████████▍| 3312/3501 [54:42<03:09,  1.00s/it]




 95%|█████████▍| 3313/3501 [54:43<03:08,  1.00s/it]




 95%|█████████▍| 3314/3501 [54:44<03:08,  1.01s/it]




 95%|█████████▍| 3315/3501 [54:45<03:07,  1.01s/it]




 95%|█████████▍| 3316/3501 [54:46<03:06,  1.01s/it]




 95%|█████████▍| 3317/3501 [54:47<03:05,  1.01s/it]




 95%|█████████▍| 3318/3501 [54:48<03:04,  1.01s/it]




 95%|█████████▍| 3319/3501 [54:49<03:02,  1.00s/it]




 95%|█████████▍| 3320/3501 [54:50<03:01,  1.00s/

we've decided to assign value -1 when it's impossible calculating distance between two nodes.
Then we'll replace this value in a reasonable way

In [114]:
shortest_path_dictionary_final

{32782: 6,
 351118: 4,
 1061148: 5,
 1061960: 4,
 1061963: 4,
 1062082: 5,
 1063756: 5,
 1065778: 5,
 216813: 5,
 350752: 5,
 350894: 4,
 350927: 5,
 351001: 5,
 351128: 4,
 543241: 3,
 1036190: 4,
 1058794: 5,
 1223412: 5,
 1394526: 3,
 430722: 4,
 1031462: 5,
 1054907: 5,
 1055543: 5,
 1056766: 6,
 1058223: 5,
 1059302: 4,
 1060200: 5,
 1060725: 5,
 1060736: 5,
 1060737: 5,
 1060739: 5,
 1060891: 5,
 1061149: 5,
 1061150: 5,
 1061664: 5,
 1061824: 4,
 1061825: 5,
 1061832: 5,
 1061848: 4,
 1061861: 4,
 1061885: 4,
 1061922: 4,
 1061939: 5,
 1062050: 5,
 1062484: 4,
 1062688: 4,
 1062748: 4,
 1064363: 5,
 1065611: 5,
 1066978: 5,
 56387: 5,
 272763: 4,
 279122: 2,
 324958: 5,
 325554: 5,
 338810: 5,
 493443: 5,
 742511: 5,
 744381: 5,
 751230: 4,
 817478: 5,
 824119: 4,
 1028775: 5,
 1028948: 5,
 1045097: 5,
 1056565: 4,
 1058108: 5,
 1058661: 5,
 1058773: 5,
 1058843: 5,
 1058933: 5,
 1059204: 5,
 1059373: 5,
 1059441: 5,
 1059645: 5,
 1059646: 5,
 1059647: 5,
 1059674: 5,
 1059741: 

Let's to check what is the maximum value of the shortest path

In [115]:
max(shortest_path_dictionary_final.values())

30

In [116]:
min(shortest_path_dictionary_final.values())

-1

In [117]:
0 in shortest_path_dictionary_final.values() 

True

In [122]:
import pickle


In [123]:
pickle.dump(shortest_path_dictionary_final, open("shortest_path_dictionary_final.p", "wb"))


In [125]:
shortest_path_dictionary_final2 = pickle.load(open("shortest_path_dictionary_final.p", "rb"))

In [126]:
type(shortest_path_dictionary_final2)

dict

The maximum value for the shortest path is 30. So we've decided to assign 100 when there insn't distance between two nodes. 
In this way we can take account of them in our analysis. 

In [128]:
for key, value in shortest_path_dictionary_final2.items():
    if shortest_path_dictionary_final2[key] == -1:
        shortest_path_dictionary_final2[key] = 100

Now we have a dictionary with the shortest distance between one category and all nodes

In [ ]:
We have to group by the distance that come from same category. At the end of this process we'll have, for each category, a dictionary.
The dictionary keys'll be the element of the category, and the value'll be the shortest path between this node and one of our input node

In [ ]:
shortest_path_1 = {}
for article in tqdm(categories_dict['English_television_actors']):
    for key, value in shortest_path_dictionary_final2.items():
        if key in categories_dict['English_television_actors']:
            shortest_path_1[key] = value

i think that this comand should work well. if it works i've thought to define a function for calculating other dictionaries